File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/Metadata.ts
module api.content {

    import MixinName = api.schema.mixin.MixinName;
    import PropertyTree = api.data.PropertyTree;
    import PropertyIdProvider = api.data.PropertyIdProvider;

    export class Metadata implements api.Cloneable, api.Equitable {

        private name: MixinName;

        private data: PropertyTree;

        constructor(name: MixinName, data: PropertyTree) {
            this.name = name;
            this.data = data;
        }

        getName(): MixinName {
            return this.name;
        }

        getData(): PropertyTree {
            return this.data;
        }

        clone(): Metadata {
            return new Metadata(this.name, this.data.copy());
        }

        equals(o: api.Equitable): boolean {
            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, Metadata)) {
                return false;
            }

            var other = <Metadata>o;

            if (!api.ObjectHelper.equals(this.name, other.name)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.data, other.data)) {
                return false;
            }

            return true;
        }

        toJson(): api.content.json.MetadataJson {
            return {
                name: this.name.toString(),
                data: this.data.toJson()
            };
        }

        static fromJson(metadataJson: api.content.json.MetadataJson, propertyIdProvider: PropertyIdProvider): Metadata {
            return new Metadata(new MixinName(metadataJson.name), PropertyTree.fromJson(metadataJson.data, propertyIdProvider));
        }

        
    }

}

File: modules/core-api/src/main/java/com/enonic/xp/core/content/Metadata.java
package com.enonic.xp.core.content;

import java.util.Objects;

import com.google.common.base.Preconditions;

import com.enonic.xp.core.data.PropertyTree;
import com.enonic.xp.core.schema.mixin.MixinName;

public final class Metadata
{
    private MixinName name;

    private PropertyTree data;

    public Metadata( final MixinName name, final PropertyTree data )
    {
        Preconditions.checkNotNull( name, "name cannot be null" );
        Preconditions.checkNotNull( data, "data cannot be null" );
        this.name = name;
        this.data = data;
    }

    public PropertyTree getData()
    {
        return data;
    }

    public void setData( final PropertyTree data )
    {
        this.data = data;
    }

    public MixinName getName()
    {
        return name;
    }

    public void setName( final MixinName name )
    {
        this.name = name;
    }

    public Metadata copy()
    {
        return new Metadata( name, data.copy() );
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final Metadata metadata = (Metadata) o;

        if ( !name.equals( metadata.name ) )
        {
            return false;
        }

        if ( !data.equals( metadata.data ) )
        {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( name, data );
    }
}
